app/controllers/supplementary/DeclarationHolderController.scala

```diff
@@ -33,29 +35,155 @@ class DeclarationHolderController @Inject()(
   appConfig: AppConfig,
   override val messagesApi: MessagesApi,
   authenticate: AuthAction,
+  errorHandler: ErrorHandler,
   customsCacheService: CustomsCacheService
 )(implicit ec: ExecutionContext)
     extends FrontendController with I18nSupport {
 
   import forms.supplementary.DeclarationHolder._
+  import forms.supplementary.DeclarationHoldersData.formId
+
+  val cacheId = appConfig.appName
 
   def displayForm(): Action[AnyContent] = authenticate.async { implicit request =>
-    customsCacheService.fetchAndGetEntry[DeclarationHolder](appConfig.appName, formId).map {
-      case Some(data) => Ok(declaration_holder(appConfig, form.fill(data)))
-      case _          => Ok(declaration_holder(appConfig, form))
+    customsCacheService.fetchAndGetEntry[DeclarationHoldersData](cacheId, formId).map {
+      case Some(data) => Ok(declaration_holder(appConfig, form, data.holders))
+      case _          => Ok(declaration_holder(appConfig, form, Seq()))
     }
   }
 
-  def saveHolderOfAuthorisation(): Action[AnyContent] = authenticate.async { implicit request =>
-    form
-      .bindFromRequest()
-      .fold(
-        (formWithErrors: Form[DeclarationHolder]) =>
-          Future.successful(BadRequest(declaration_holder(appConfig, formWithErrors))),
-        form =>
-          customsCacheService.cache[DeclarationHolder](appConfig.appName, formId, form).map { _ =>
-            Redirect(controllers.supplementary.routes.DestinationCountriesController.displayForm())
+  def submitHoldersOfAuthorisation(): Action[AnyContent] = authenticate.async { implicit request =>
+    val boundForm = form.bindFromRequest()
+
+    val actionType =
+      request.body.asFormUrlEncoded.flatMap(_.get("action")).flatMap(_.headOption).getOrElse("Wrong action")
+
+    val cachedData = customsCacheService
+      .fetchAndGetEntry[DeclarationHoldersData](cacheId, formId)
+      .map(_.getOrElse(DeclarationHoldersData(Seq())))
+
+    cachedData.flatMap { cache =>
+      boundForm
+        .fold(
+          (formWithErrors: Form[DeclarationHolder]) =>
+            Future.successful(BadRequest(declaration_holder(appConfig, formWithErrors, cache.holders))),
+          validForm =>
+            actionType match {
+              case "Add"                             => addAnotherHolderHandler(validForm, cache)
+              case "Save and continue"               => saveAndContinueHandler(validForm, cache)
+              case value if value.contains("Remove") => removeHolderHandler(retrieveHolder(actionType), cache)
+              case _                                 => errorHandler.displayErrorPage()
+          }
+        )
+    }
+  }
+
+  private def addAnotherHolderHandler(
```

Comment:
I'd suggest a different names for methods. Handler is better for a class but here it is not required to be readable. 'addHolder' should be enough.